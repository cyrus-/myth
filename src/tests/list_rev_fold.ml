let s = "
type nat =
  | O
  | S of nat

type list =
  | Nil
  | Cons of nat * list

let rec fold (l:list) (f:list -> nat -> list) (acc:list) : list =
  match l with
  | Nil -> acc
  | Cons (x, l) -> fold l f (f acc x)
;;


let snoc : list -> nat -> list =
  let rec f (l:list) : nat -> list =
    fun (n:nat) ->
      match l with
      | Nil -> Cons (n, Nil)
      | Cons (x, xs) -> Cons (x, f xs n)
  in
    f
;;

let list_rev_fold : list -> list |>
  { [] => []
  | [0] => [0]
  | [1] => [1]
  | [0;1] => [1;0]
  | [0;0;1] => [1;0;0]
  } = ?"
