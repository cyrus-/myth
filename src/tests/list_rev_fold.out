(* Generated Thu Nov 13 00:35:51 2014 *)
let list_rev_fold : list -> list =
  fun (l1:list) ->
    fold l1 (fun (l1:list) -> fun (n1:nat) -> Cons (n1, l1)) Nil
;;
